File: LeetCode/0250.Count-Univalue-Subtrees/Count-Univalue-Subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int res;
    public int countUnivalSubtrees(TreeNode root) {
        res = 0;
        helper(root);
        return res;
    }
    public boolean helper(TreeNode root){
        if(root == null) return true;
        boolean left = helper(root.left);
        boolean right = helper(root.right);
        if(left && right){
            if(root.left != null && root.val != root.left.val){
                return false;
            }
            if(root.right != null && root.val != root.right.val){
                return false;
            }
            res++;
            return true;
        }
        return false;
    }
}

File: LeetCode/0250.Count-Univalue-Subtrees/Count-Univalue-Subtrees.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def countUnivalSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(node):
            if not node: return (True, 0)
            isleft, left = dfs(node.left)
            isright, right = dfs(node.right)
            if isleft and (not node.left or node.left.val == node.val) and isright and (not node.right or node.right.val == node.val):
                return (True, left + right + 1)
            return (False, left + right)
        
        return dfs(root)[1]
